mm2src/mm2_main/src/rpc/lp_commands/lr_swap/lr_impl.rs

```diff
@@ -0,0 +1,350 @@
+//! Finding best quote to do swaps with liquidity routing (LR) support
+//! Swaps with LR run additional interim swaps in EVM chains to convert one token into another token suitable to do a normal atomic swap.
+
+use crate::lp_ordermatch::RpcOrderbookEntryV2;
+use crate::rpc::lp_commands::one_inch::errors::ApiIntegrationRpcError;
+use crate::rpc::lp_commands::one_inch::rpcs::get_coin_for_one_inch;
+use coins::eth::{u256_to_big_decimal, wei_from_big_decimal};
+use coins::hd_wallet::DisplayAddress;
+use coins::lp_coinfind_or_err;
+use coins::MmCoin;
+use coins::NumConversResult;
+use coins::Ticker;
+use common::log;
+use ethereum_types::Address as EthAddress;
+use ethereum_types::{FromDecStrErr, U256};
+use futures::future::join_all;
+use mm2_core::mm_ctx::MmArc;
+use mm2_err_handle::prelude::*;
+use mm2_number::MmNumber;
+use num_traits::CheckedDiv;
+use std::collections::HashMap;
+use trading_api::one_inch_api::classic_swap_types::{ClassicSwapData, ClassicSwapQuoteParams};
+use trading_api::one_inch_api::client::{ApiClient, PortfolioApiMethods, PortfolioUrlBuilder, SwapApiMethods,
+                                        SwapUrlBuilder};
+use trading_api::one_inch_api::portfolio_types::{CrossPriceParams, CrossPricesSeries, DataGranularity};
+
+/// To estimate src/dst price query price history for last 5 min
+const CROSS_PRICES_GRANULARITY: DataGranularity = DataGranularity::FiveMin;
+/// Use no more than 1 price history samples to estimate src/dst price
+const CROSS_PRICES_LIMIT: u32 = 1;
+
+#[inline]
+fn mm_number_to_u256(mm_number: &MmNumber) -> Result<U256, FromDecStrErr> {
+    U256::from_dec_str(mm_number.to_ratio().to_integer().to_string().as_str())
+}
+
+#[inline]
+fn mm_number_from_u256(u256: U256) -> MmNumber { MmNumber::from(u256.to_string().as_str()) }
+
+#[inline]
+fn wei_from_coins_mm_number(mm_number: &MmNumber, decimals: u8) -> NumConversResult<U256> {
+    wei_from_big_decimal(&mm_number.to_decimal(), decimals)
+}
+
+#[inline]
+#[allow(unused)]
+fn wei_to_coins_mm_number(u256: U256, decimals: u8) -> NumConversResult<MmNumber> {
+    Ok(MmNumber::from(u256_to_big_decimal(u256, decimals)?))
+}
+
+/// Internal struct to collect data for selecting the best swap with LR
+struct LrData {
+    /// Order to fill
+    order: RpcOrderbookEntryV2,
```

Comment:
we store a copy of the whole order struct but i see we only use the `price` and `coin` fields of it. with `coin` field also being included in the key of the `LrDataMap` already.